### ISSUE TITLE
uniqExact using sip128 code  may have collision 

### ISSUE THREAD

**Issue by templarzq** [ClickHouse/ClickHouse#9506]
version: v19.16  / master

sql:   select uniqExact(user_id) from t;

comment:  user_id  column type:  String
source code:

dbms/src/AggregateFunctions/AggregateFunctionUniq.h

/// For rows, we put the SipHash values (128 bits) into the hash table.
template <>
struct AggregateFunctionUniqExactData<String>
{
    using Key = UInt128;

    /// When creating, the hash table must be small.
    using Set = HashSet<
        Key,
        UInt128TrivialHash,
        HashTableGrower<3>,
        HashTableAllocatorWithStackMemory<sizeof(Key) * (1 << 3)>>;

    Set set;

    static String getName() { return "uniqExact"; }
};

when column type is String , uniqExact use uint128 sip128hash code set instead of  original string .
what if  the hash code collision happened?

**Comment by templarzq** [comment]
Does this lead to fault uniq result?

**Comment by alexey-milovidov** [comment]
> What if the hash code collision happened?

You write an article about the first found collision in SipHash128.

We know in advance that there are collisions in SipHash128 and the probabilities of collisions with random values are:
- that the random value will collide with the fixed value S: 1 / 2 ^ 128;
- that there are at least one collision in a set of random values of size N: approx 1 / 2 for N = 2 ^ 64 (birthday principle).

SipHash128 is considered cryptographically secure. It means that there is no better way to search for collisions than brute force.

Three considerations exist:
1. Even if there are no exploitable weakness found in SipHash, they can be found in future, as it already happened with MD5, SHA1, etc... Then we will change the hash function or change the algorithm. Even if the weakness in SipHash will be found, it will not compromise the quality of the algorithm for non malicious user.
2. We use SipHash 2-4 variant, although SipHash 4-8 is recommended for better security (but also note that we are using 128 bit variant instead of 64 bit).
3. You can still find a collision with brute force, it will need about hundred exabytes of random access storage - this is doable with current level of technology but is very expensive.

TLDR: You don't have to worry.

**Comment by templarzq** [comment]
> > What if the hash code collision happened?
> 
> You write an article about the first found collision in SipHash128.
> 
> We know in advance that there are collisions in SipHash128 and the probabilities of collisions with random values are:
> 
> * that the random value will collide with the fixed value S: 1 / 2 ^ 128;
> * that there are at least one collision in a set of random values of size N: approx 1 / 2 for N = 2 ^ 64 (birthday principle).
> 
> SipHash128 is considered cryptographically secure (as of Mar 2020). It means that there is no better way to search for collisions than brute force.
> 
> Three considerations exist:
> 
> 1. Even if there are no exploitable weakness found in SipHash, they can be found in future, as it already happened with MD5, SHA1, etc... Then we will change the hash function or change the algorithm. Even if the weakness in SipHash will be found, it will not compromise the quality of the algorithm for non malicious user.
> 2. We use SipHash 2-4 variant, although SipHash 4-8 is recommended for better security (but also note that we are using 128 bit variant instead of 64 bit).
> 3. You can still find a collision with brute force, it will need about hundred exabytes of random access storage - this is doable with current level of technology but is very expensive.
> 
> TLDR: You don't have to worry.
> 
> > Does this lead to fault uniq result?
> 
> No.

So what if we use sip64 instead of sip128?

**Comment by templarzq** [comment]
any way,i'm not worried about brute force attack, just to predict that if there are collision occured in a eb/tb string type column .

**Comment by alexey-milovidov** [comment]
> So what if we use sip64 instead of sip128?

Then result will be affected by collisions and you will certainly get a few collisions on a set of size about 10 billion. For larger sets, result will be worse.

**Comment by templarzq** [comment]
is there any essential difference between sip64/sip128?

**Comment by alexey-milovidov** [comment]
Yes.

**Comment by templarzq** [comment]
so if 

> > So what if we use sip64 instead of sip128?
> 
> Then result will be affected by collisions and you will certainly get a few collisions on a set of size about 10 billion. For larger sets, result will be worse.

so if sip64 collision exists, how to ensure that sip128 has no collison in very large datasets?

**Comment by alexey-milovidov** [comment]
Both SipHash64 and SipHash128 have collisions, but to get the collision of 128bit variant you have to observe quadratic amount of values comparing to SipHash64.

**Comment by templarzq** [comment]
what about to stor original string value in the hashset insead of sip128 code? because in normal way string value length is smaller than 16byte(128bit code)

**Comment by amosbird** [comment]
I understand the concern of @templarzq (I was there before). I'll post the chat history with @alexey-milovidov here and hope it would help clarify this idea.

```
Amos Bird, [14.03.19 09:34]
I've also noticed that we use the cryptic hash functions for giant keys instead of pointers ... not sure if that's really valid. someday we might hit a collision

Alexey Milovidov, [14.03.19 09:37]
128bit cryptographic hash functions - will hit a collision with reasonable probability for sets of size about 2^64. This is Ok for almost all applications. I cannot imagine applications when this is not satisfactory.

Amos Bird, [14.03.19 09:38]
yeah, but that's probability. do we add randomization to those hash functions?

Alexey Milovidov, [14.03.19 09:39]
No.

Amos Bird, [14.03.19 09:39]
I rememeber there is a list of string (human readable, meaningful) that deterministicly collide

Alexey Milovidov, [14.03.19 09:41]
Yes. But you cannot find this set for SipHash128.
(If you can - you should write a paper about weakness of SipHash.)

Alexey Milovidov, [14.03.19 09:43]
(No one even has found a pair of strings that collide.)

Amos Bird, [14.03.19 09:45]
[In reply to Alexey Milovidov]
so is that a well-known way of doing aggregations using cryptographic hash functions?

Alexey Milovidov, [14.03.19 09:50]
[In reply to Amos Bird]
I'm not sure that it's well known.
Someone may argue that 128 bit is not enough... but it's questionable.

For example, git uses 160bit SHA-1 hash function. SHA-1 hash known weaknesses and is considered obsolete. Collision pairs has been found. You can break git and svn with uploading some specially prepared files.

https://arstechnica.com/information-technology/2017/02/at-deaths-door-for-years-widely-used-sha1-function-is-now-dead/

For SipHash, there is no known collisions. If there will be found (or near to be found), we are going to change the hash function.

Alexey Milovidov, [14.03.19 09:51]
PS. For SipHash64 it's easy to find collisions. For 128bit - it isn't.

Amos Bird, [14.03.19 09:52]
> I'm not sure that it's well known.

this is the kind of tricks that combines theorical and practical, very cool

Amos Bird, [14.03.19 09:55]
so why not using siphash for StringRef?

Amos Bird, [14.03.19 09:55]
we don't need to compare strings anymore

Alexey Milovidov, [14.03.19 09:58]
String comparison (with calculating non-crypto hash for hash table) appeared to be faster than calculating crypto hash.

Amos Bird, [14.03.19 09:59]
I just realised that if , by using clickhouse, we find a collision of siphash, it's a win-win for both world. So the choice of doing this is really a smart move. In fact, we should even propagate this idea so more systems use this to help testing the cryptograph functions validity

Alexey Milovidov, [14.03.19 09:59]
Actually, we don't use SipHash for GROUP BY, because "serialized" method appeared to be faster. We only use for Sets and Joins and arrayEnumerateUniq/etc.

Amos Bird, [14.03.19 09:59]
[In reply to Alexey Milovidov]
hmm, I thought SipHash is somehow faster

Alexey Milovidov, [14.03.19 10:00]
[In reply to Amos Bird]
But collision in SipHash will most likely to be unnoticed.

Amos Bird, [14.03.19 10:00]
[In reply to Alexey Milovidov]
serailized? You mean toString?

Alexey Milovidov, [14.03.19 10:01]
AggregationMethodSerialized

Alexey Milovidov, [14.03.19 10:02]
There was "AggregationMethodHashed" but it was removed because it was slower.

Amos Bird, [14.03.19 10:03]
does AggregationMethodSerialized store keys as flat storage or pointers?

Alexey Milovidov, [14.03.19 10:03]
Pointers (StringRef)

Amos Bird, [14.03.19 10:03]
ah, ok
```

**Comment by amosbird** [comment]
Anyway, if one day you found a collision when using `uniqExact`, try writing a blog and get famous in cryptographic communities.

**Comment by templarzq** [comment]
LOL

**Comment by blinkov** [comment]
@templarzq, do you have any further questions?

**Comment by templarzq** [comment]
no
